lib/services/role.go

```diff
@@ -201,10 +179,91 @@ func RoleForUser(u User) Role {
 				AppLabels:        Labels{Wildcard: []string{Wildcard}},
 				KubernetesLabels: Labels{Wildcard: []string{Wildcard}},
 				DatabaseLabels:   Labels{Wildcard: []string{Wildcard}},
-				Rules:            CopyRulesSlice(AdminUserRules),
+				Rules: []Rule{
+					NewRule(KindRole, RW()),
+					NewRule(KindAuthConnector, RW()),
+					NewRule(KindSession, RO()),
+					NewRule(KindTrustedCluster, RW()),
+					NewRule(KindEvent, RO()),
+				},
+			},
+		},
+	}
+}
+
+// NewOSSUserRole is a role for enabling RBAC for open source users.
+// This is a limited role
+func NewOSSUserRole(name ...string) Role {
+	role := &RoleV3{
+		Kind:    KindRole,
+		Version: V3,
+		Metadata: Metadata{
+			Name:      teleport.OSSUserRoleName,
+			Namespace: defaults.Namespace,
+		},
+		Spec: RoleSpecV3{
+			Options: RoleOptions{
+				CertificateFormat: teleport.CertificateFormatStandard,
+				MaxSessionTTL:     NewDuration(defaults.MaxCertDuration),
+				PortForwarding:    NewBoolOption(true),
+				ForwardAgent:      NewBool(true),
+				BPF:               defaults.EnhancedEvents(),
+			},
+			Allow: RoleConditions{
+				Namespaces:       []string{defaults.Namespace},
+				NodeLabels:       Labels{Wildcard: []string{Wildcard}},
+				AppLabels:        Labels{Wildcard: []string{Wildcard}},
+				KubernetesLabels: Labels{Wildcard: []string{Wildcard}},
+				DatabaseLabels:   Labels{Wildcard: []string{Wildcard}},
+				DatabaseNames:    []string{teleport.TraitInternalDBNamesVariable},
+				DatabaseUsers:    []string{teleport.TraitInternalDBUsersVariable},
+				Rules: []Rule{
+					NewRule(KindEvent, RO()),
+				},
 			},
 		},
 	}
+	role.SetLogins(Allow, []string{teleport.TraitInternalLoginsVariable})
+	role.SetKubeUsers(Allow, []string{teleport.TraitInternalKubeUsersVariable})
+	role.SetKubeGroups(Allow, []string{teleport.TraitInternalKubeGroupsVariable})
+	return role
+}
+
+// NewOSSGithubRole creates a role for enabling RBAC for open source Github users
+func NewOSSGithubRole(logins []string, kubeUsers []string, kubeGroups []string) Role {
```

Comment:
Can `NewAdminRole`, `RoleForUser`, `NewOSSUserRole`, and `NewOSSGithubRole` be consolidated? Looks like the bulk is the same with some slight variations.

Looks like there is already some (unintentional?) drift because `RoleForUser` is missing `DatabaseNames` and `DatabaseUsers` which the others all have.

Reply:
@russjones those are slightly different roles,

RoleForUser is used for tests only, NewOSSUserRole is not privileged any more, NewOSSGithubRole adds extra custom traits, so would be hard to consolidate those

Reply:
I was thinking more along the lines how we copy admin rules `CopyRulesSlice(ExtendedAdminUserRules)` so something like the below. Reduces chances of someone only fixing something in one place or roles being different enough during tests that we don't capture what would happen in produciton.

```go
// Role is filled out with common defaults.
role := &RoleV3{
	Kind:    KindRole,
	Version: V3,
	Metadata: Metadata{
		Name:      teleport.OSSUserRoleName,
		Namespace: defaults.Namespace,
	},
	Spec: RoleSpecV3{
		Options: copyOptions(defaultOptions),
		Allow:   copyConditions(defaultConditions),
	}
}
// Override any values needed here.
role.SetLogins(Allow, []string{teleport.TraitInternalLoginsVariable})
role.SetKubeUsers(Allow, []string{teleport.TraitInternalKubeUsersVariable})
role.SetKubeGroups(Allow, []string{teleport.TraitInternalKubeGroupsVariable})
```